tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,252 @@
+    for (Tensor t : tensors) {
+      detach(t);
+    }
+  }
+
+  /**
+   * @see #detach(Tensor)
+   */
+  public static <T extends TensorContainer> T detach(T tensors) {
+    detach(tensors.tensors());
+    return tensors;
+  }
+
+  /**
+   * @see #detach(Tensor)
+   */
+  public static void detach(TensorContainer... tensors) {
+    for (TensorContainer ht : tensors) {
+      detach(ht);
+    }
+  }
+
+  /**
+   * @see #detach(Tensor)
+   */
+  public static <T extends Iterable<? extends Tensor>> T detach(T tensors) {
+    tensors.forEach(TensorScope::detach);
+    return tensors;
+  }
+
+  /**
+   * @see #detach(Tensor)
+   */
+  @SafeVarargs
+  public static void detach(Iterable<? extends Tensor>... tensors) {
+    for (Iterable<? extends Tensor> iterable : tensors) {
+      detach(iterable);
+    }
+  }
+
+  /**
+   * Attach a tensor to this scope.  This happens automatically to tensors that are created in the scope.
+   *
+   * @return this
+   */
+  public <T extends Tensor> T attach(T tensor) {
+    if (this.closed) {
+      throw new IllegalStateException("Scope has been closed, can not attach new tensor.");
+    }
+
+    RawTensor rt = tensor.asRawTensor();
+    detach(tensor);
+    rt.tensorScope = this;
+    tensors.add(rt);
+
+    return tensor;
+  }
+
+  /**
+   * @see #attach(Tensor)
+   */
+  public void attach(Tensor... tensors) {
+    if (tensors != null) {
+      for (Tensor t : tensors) {
+        attach(t);
+      }
+    }
+  }
+
+  /**
+   * @see #attach(Tensor)
+   */
+  public <T extends TensorContainer> T attach(T tensors) {
+    attach(tensors.tensors());
+    return tensors;
+  }
+
+  /**
+   * @see #attach(Tensor)
+   */
+  public void attach(TensorContainer... tensors) {
+    if (tensors != null) {
+      for (TensorContainer ht : tensors) {
+        attach(ht);
+      }
+    }
+  }
+
+  /**
+   * @see #attach(Tensor)
+   */
+  public <T extends Iterable<? extends Tensor>> T attach(T tensors) {
+    tensors.forEach(this::attach);
+
+    return tensors;
+  }
+
+  /**
+   * @see #attach(Tensor)
+   */
+  @SafeVarargs
+  public final void attach(Iterable<? extends Tensor>... tensors) {
+    if (tensors != null) {
+      for (Iterable<? extends Tensor> ht : tensors) {
+        attach(ht);
+      }
+    }
+  }
+
+  /**
+   * @see #attach(Tensor)
+   */
+  public TensorScope withTensors(Tensor... tensors) {
+    attach(tensors);
+    return this;
+  }
+
+  /**
+   * @see #attach(Tensor)
+   */
+  public TensorScope withTensors(TensorContainer... tensors) {
+    attach(tensors);
+    return this;
+  }
+
+  /**
+   * @see #attach(Tensor)
+   */
+  @SafeVarargs
+  public final TensorScope withTensors(Iterable<? extends Tensor>... tensors) {
+    attach(tensors);
+    return this;
+  }
+
+  /**
+   * Gets whether the scope is closed.
+   */
+  public boolean isClosed() {
+    return closed;
+  }
+
+  private boolean closed = false;
+  private final Set<RawTensor> tensors = Collections.newSetFromMap(new WeakHashMap<>());
```

Comment:
By relying on weak references, we can end up deallocating prematurely, like I point out in issue https://github.com/tensorflow/java/issues/208#issuecomment-785552930.
Is there a way to guarantee that TF Core never keeps references to tensors on its own? And is this something we really want to enforce?

Reply:
The difference from `PointerScope` is that we're matching Java semantics, not C++s, so assuming a unreachable object is GC'd/not accessible is entirely correct.  For Java, `keepReferenceTo` wouldn't be able to (reasonably) expect arguments it doesn't hold onto itself to be held on to.  `TensorScope` only provides the guarantee that tensors will be closed when the scope closed or sooner, we don't want it to keep things being collected by JavaCPP's cleaner if they are unreachable earlier than the end of the scope.

Since we're calling TF Core's deallocator method to do the deallocation, I would consider any issues with it holding references to be a bug on their side, and unlikely.  In theory, even if it does hold onto references, it should handle that in the deallocation method we call (i.e. with some kind of internal reference counting).

Reply:
In any case, I don't think it's a good idea to rely on the GC, since it may never deallocate anything at all. I'll leave it up to others to decide on whether we really want to keep doing things this way though.

Reply:
I agree, but this isn't relying on GC, it's just not preventing it from being used.

Reply:
"Sometimes it works, sometimes it doesn't work", that's not a great pitch line :) For users that don't mind, we may want to make TensorScope optional, but we should provide a way to make it work reliably, always.

Reply:
No, what I mean is that tensor scope's contract is "ensures all tensors are closed when the scope closes".  It always works, regardless of GC use.  What this does is it allows tensors to be closed by GC earlier.  That doesn't break the contract.

Reply:
Right, but what is that useful for, exactly? Say you have some piece of code that works fine on your development machine with small datasets, where the GC gets fired often enough, no problem, and then you load that up in production, where the heap is full of other crap from other frameworks and the CPU is busy doing something else, so the GC ends up running a lot less frequently, and, boom, it runs out of memory. What's the point of using TensorScope if not to have some of guarantee that things are going to work like the developer think they are going to work?

Reply:
On my side, like I've mentioned in https://github.com/tensorflow/java/issues/208#issuecomment-787000189, I'm comfortable with the idea of keeping only weak references to tensors (or any native resource) in the scopes for not preventing the GC to free their memory if it has a chance to do it, which does more good than harm. I'm pretty confident than resources allocated via the C API are ultimately copied by the TF runtime if it really needs them to remain alive.